File: crates/vdb-node/src/filter.rs
//! Turning a JavaScript object into a filter.
//!
//! Node binds the engine directly rather than through the C ABI, so it does not need the
//! postfix builder the C bindings use — it can take the shape a JavaScript developer would
//! write anyway:
//!
//! ```js
//! { category: 'tools', price: { $lt: 50 } }              // both must hold
//! { $or: [{ category: 'toys' }, { price: { $gt: 50 } }] }
//! { $not: { archived: true } }
//! { tags: { $contains: 'sharp' } }
//! { price: { $exists: true } }
//! ```
//!
//! The query-object convention is borrowed deliberately: it is what a JavaScript developer
//! expects from a database, and inventing a different one would cost familiarity for nothing.
//! What is *not* borrowed is MongoDB's semantics — the type rules are vdb's, and they are total.
//! `docs/api/filters.md` is the reference, including the three rules that surprise people.
//!
//! # Bare keys mean equality, and several keys mean conjunction
//!
//! `{ a: 1, b: 2 }` is `a == 1 && b == 2`. That is what the shape looks like it means, and a
//! filter language whose obvious reading is wrong is worse than one with no shorthand at all.

use napi::bindgen_prelude::{Object, Result};
use napi::{Error as NapiError, JsString, JsUnknown, ValueType};

use vdb_core::filter::Filter;
use vdb_core::metadata::Value;

/// Deepest nesting accepted, matching the engine's own limit.
///
/// Checked on the way down rather than after building: this walk recurses, and a deeply nested
/// object from an untrusted source would otherwise overflow the stack, which aborts rather than
/// returning an error anyone can handle.
const MAX_DEPTH: usize = 32;

/// Parse a filter object.
///
/// No `Env` parameter: `Object` carries its own, and taking one changed how napi-rs mapped the
/// JavaScript arguments — the filter silently arrived as `None` and every filter was ignored.
pub(crate) fn parse(object: Object) -> Result<Filter> {
    parse_object(object, 1)
}

fn parse_object(object: Object, depth: usize) -> Result<Filter> {
    if depth > MAX_DEPTH {
        return Err(NapiError::from_reason(format!(
            "filter nested deeper than {MAX_DEPTH}"
        )));
    }
    let names = object.get_property_names()?;
    let count = names.get_array_length()?;
    let mut clauses = Vec::with_capacity(count as usize);

    for index in 0..count {
        let key = property_name(&names, index)?;
        let value: JsUnknown = object.get_named_property(&key)?;

        let clause = match key.as_str() {
            "$and" => Filter::all(parse_list(value, depth)?),
            "$or" => Filter::any(parse_list(value, depth)?),
            "$not" => {
                if value.get_type()? != ValueType::Object {
                    return Err(NapiError::from_reason("$not takes a filter object"));
                }
                // SAFETY: the type was checked immediately above.
                let inner: Object = unsafe { value.cast() };
                Filter::negate(parse_object(inner, depth + 1)?)
            }
            other if other.starts_with('$') => {
                return Err(NapiError::from_reason(format!(
                    "unknown filter operator {other:?}; expected $and, $or or $not at the top level"
                )))
            }
            field => parse_field(field, value, depth)?,
        };
        clauses.push(clause);
    }

    // Several keys in one object mean conjunction; one means itself. An empty object matches
    // everything, which is the identity of `and` and what `{}` visibly ought to mean.
    Ok(match clauses.len() {
        1 => clauses
            .into_iter()
            .next()
            .unwrap_or_else(|| Filter::all(vec![])),
        _ => Filter::all(clauses),
    })
}

/// One property name, as a Rust string.
///
/// Property names arrive as JavaScript strings rather than Rust ones; the conversion is here
/// so the two walkers below do not each spell it out.
fn property_name(names: &Object, index: u32) -> Result<String> {
    let raw: JsString = names.get_element(index)?;
    raw.into_utf8()?.into_owned()
}

fn parse_list(value: JsUnknown, depth: usize) -> Result<Vec<Filter>> {
    if value.get_type()? != ValueType::Object {
        return Err(NapiError::from_reason(
            "$and and $or take an array of filters",
        ));
    }
    // SAFETY: the type was checked immediately above.
    let array: Object = unsafe { value.cast() };
    if !array.is_array()? {
        return Err(NapiError::from_reason(
            "$and and $or take an array of filters",
        ));
    }
    let length = array.get_array_length()?;
    let mut out = Vec::with_capacity(length as usize);
    for index in 0..length {
        let element: Object = array.get_element(index)?;
        out.push(parse_object(element, depth + 1)?);
    }
    Ok(out)
}

/// One field's clause: a bare value means equality, an object means operators.
fn parse_field(field: &str, value: JsUnknown, depth: usize) -> Result<Filter> {
    if value.get_type()? == ValueType::Object {
        // SAFETY: the type was checked immediately above.
        let operators: Object = unsafe { value.cast() };
        // An array as a field value would be ambiguous — equality with the array, or a set
        // membership test? — so it is refused rather than guessed at.
        if operators.is_array()? {
            return Err(NapiError::from_reason(format!(
                "field {field:?}: an array is not a filter value; use {{ $contains: … }} for \
                 array membership"
            )));
        }
        return parse_operators(field, operators, depth);
    }
    Ok(Filter::eq(field, scalar(value, field)?))
}

fn parse_operators(field: &str, operators: Object, depth: usize) -> Result<Filter> {
    let _ = depth;
    let names = operators.get_property_names()?;
    let count = names.get_array_length()?;
    let mut clauses = Vec::with_capacity(count as usize);

    for index in 0..count {
        let op = property_name(&names, index)?;
        let raw: JsUnknown = operators.get_named_property(&op)?;

        let clause = match op.as_str() {
            "$eq" => Filter::eq(field, scalar(raw, field)?),
            "$ne" => Filter::ne(field, scalar(raw, field)?),
            "$gt" => Filter::gt(field, scalar(raw, field)?),
            "$gte" => Filter::gte(field, scalar(raw, field)?),
            "$lt" => Filter::lt(field, scalar(raw, field)?),
            "$lte" => Filter::lte(field, scalar(raw, field)?),
            "$contains" => Filter::contains(field, scalar(raw, field)?),
            "$startsWith" => match scalar(raw, field)? {
                Value::Str(prefix) => Filter::starts_with(field, prefix),
                _ => {
                    return Err(NapiError::from_reason(format!(
                        "field {field:?}: $startsWith takes a string"
                    )))
                }
            },
            "$in" => Filter::in_values(field, scalar_list(raw, field)?),
            "$nin" => Filter::not_in(field, scalar_list(raw, field)?),
            "$exists" => {
                let present: bool = raw.coerce_to_bool()?.get_value()?;
                if present {
                    Filter::exists(field)
                } else {
                    Filter::is_null(field)
                }
            }
            other => {
                return Err(NapiError::from_reason(format!(
                    "field {field:?}: unknown operator {other:?}"
                )))
            }
        };
        clauses.push(clause);
    }

    Ok(match clauses.len() {
        1 => clauses
            .into_iter()
            .next()
            .unwrap_or_else(|| Filter::all(vec![])),
        _ => Filter::all(clauses),
    })
}

fn scalar_list(value: JsUnknown, field: &str) -> Result<Vec<Value>> {
    if value.get_type()? != ValueType::Object {
        return Err(NapiError::from_reason(format!(
            "field {field:?}: $in and $nin take an array"
        )));
    }
    // SAFETY: the type was checked immediately above.
    let array: Object = unsafe { value.cast() };
    if !array.is_array()? {
        return Err(NapiError::from_reason(format!(
            "field {field:?}: $in and $nin take an array"
        )));
    }
    let length = array.get_array_length()?;
    let mut out = Vec::with_capacity(length as usize);
    for index in 0..length {
        let element: JsUnknown = array.get_element(index)?;
        out.push(scalar(element, field)?);
    }
    Ok(out)
}

/// Convert one JavaScript value.
fn scalar(value: JsUnknown, field: &str) -> Result<Value> {
    Ok(match value.get_type()? {
        ValueType::String => Value::Str(value.coerce_to_string()?.into_utf8()?.into_owned()?),
        ValueType::Boolean => Value::Bool(value.coerce_to_bool()?.get_value()?),
        ValueType::Null | ValueType::Undefined => Value::Null,
        ValueType::Number => {
            let n: f64 = value.coerce_to_number()?.get_double()?;
            // Integral numbers become integers, matching how metadata is stored — otherwise
            // `{ count: 3 }` would compare a float against a stored integer and, although the
            // engine handles that correctly, the asymmetry would be a trap waiting for the day
            // it does not.
            if n.fract() == 0.0 && n.abs() < 9e15 {
                Value::I64(n as i64)
            } else {
                Value::F64(n)
            }
        }
        other => {
            return Err(NapiError::from_reason(format!(
                "field {field:?}: filter values must be strings, numbers, booleans or null; \
                 got {other:?}"
            )))
        }
    })
}

File: sdk/react-native/src/filter.js
// Turning a filter query object into the postfix sequence the C ABI takes.
//
// A filter is a tree, and C has no good way to receive one, so `vdb.h` takes a stack: push
// leaves, then combine the top N. Node binds the engine directly and can hand it a tree, but
// everything reaching the engine through the C ABI — this SDK, Swift, Java — has to flatten one.
//
// This is done in JavaScript rather than in C++ for the same reason the bridge holds the rest of
// the logic: it is the half that can be tested without a device. `test/filter.test.js` runs the
// whole of it in Node, and `cpp/vdb_bridge.cpp` replays whatever it produces onto a real
// `vdb_filter_t` and checks the answers against the real engine.
//
// The query-object convention is borrowed from `@isais-logic/isha-vector-db-node` deliberately, and the
// semantics are the engine's own — `docs/api/filters.md` is the reference. Divergence between
// SDKs is the fastest way to make a cross-platform library untrustworthy, so where this file has
// a choice it makes the same one `crates/isha-vector-db-node/src/filter.rs` makes.

import { VdbError } from './error.js';

/** Step kinds, matching `MetaField`/`FilterOp` in `cpp/vdb_bridge.h`. */
const STEP = {
  compareString: 1,
  compareI64: 2,
  compareF64: 3,
  compareBool: 4,
  unary: 5,
  combine: 6,
};

/** `vdb_op_t`. */
const OP = {
  eq: 1,
  ne: 2,
  gt: 3,
  gte: 4,
  lt: 5,
  lte: 6,
  startsWith: 7,
  contains: 8,
};

/** `vdb_unary_t`. */
const UNARY = { exists: 1, isNull: 2 };

/** `vdb_combine_t`. */
const COMBINE = { and: 1, or: 2, not: 3 };

/**
 * Deepest nesting accepted, matching the engine's own limit and Node's.
 *
 * Checked on the way down rather than after building: this walk recurses, and a deeply nested
 * object from an untrusted source would otherwise overflow the stack, which kills the app rather
 * than producing an error anyone can handle.
 */
const MAX_DEPTH = 32;

/** One step, with every property present so the JSI layer reads a fixed shape and never branches. */
function step(kind, { field = '', op = 0, text = '', number = 0, flag = false, count = 0 } = {}) {
  return { kind, field, op, text, number, flag, count };
}

function fail(message) {
  // `INVALID_ARGUMENT`. The engine would say the same thing about the same mistake; catching it
  // here only means it is said before a native call rather than after one.
  return new VdbError(1002, message);
}

/**
 * Compile a filter object into a postfix step list.
 *
 * `undefined` and `{}` both compile to an empty list, which the bridge reads as "no filter" —
 * `{}` is the identity of `and`, and searching everything is what it visibly ought to mean.
 */
export function compileFilter(filter) {
  if (filter === undefined || filter === null) return [];
  if (typeof filter !== 'object' || Array.isArray(filter)) {
    throw fail('a filter must be an object');
  }
  const steps = [];
  const clauses = walkObject(filter, steps, 1);
  // A filter of no clauses is not pushed at all, so an empty object stays an empty list rather
  // than becoming an `and` of nothing that the stack would then have to hold.
  if (clauses === 0) return [];
  conjoin(steps, clauses);
  return steps;
}

/** Combine `count` expressions already on the stack, unless there is only the one. */
function conjoin(steps, count) {
  if (count > 1) steps.push(step(STEP.combine, { op: COMBINE.and, count }));
}

/**
 * Walk one filter object, pushing its clauses. Returns how many it pushed.
 *
 * Several keys in one object mean conjunction and a bare value means equality — which is what
 * the shape looks like it means, and a filter language whose obvious reading is wrong is worse
 * than one with no shorthand at all.
 */
function walkObject(object, steps, depth) {
  if (depth > MAX_DEPTH) throw fail(`filter nested deeper than ${MAX_DEPTH}`);
  let clauses = 0;

  for (const [key, value] of Object.entries(object)) {
    if (value === undefined) continue;
    switch (key) {
      case '$and':
      case '$or': {
        const list = requireFilterList(key, value);
        // An empty `$and` matches everything and an empty `$or` matches nothing. Neither can be
        // expressed as a combine of zero, so they are refused rather than silently becoming the
        // other one.
        if (list.length === 0) throw fail(`${key} needs at least one filter`);
        for (const each of list) {
          const pushed = walkObject(requireObject(key, each), steps, depth + 1);
          if (pushed === 0) throw fail(`${key} cannot contain an empty filter`);
          conjoin(steps, pushed);
        }
        if (list.length > 1) {
          steps.push(step(STEP.combine, {
            op: key === '$and' ? COMBINE.and : COMBINE.or,
            count: list.length,
          }));
        }
        clauses++;
        break;
      }
      case '$not': {
        const pushed = walkObject(requireObject('$not', value), steps, depth + 1);
        if (pushed === 0) throw fail('$not cannot negate an empty filter');
        conjoin(steps, pushed);
        steps.push(step(STEP.combine, { op: COMBINE.not, count: 1 }));
        clauses++;
        break;
      }
      default: {
        if (key.startsWith('$')) {
          throw fail(
            `unknown filter operator ${JSON.stringify(key)}; expected $and, $or or $not at the ` +
              'top level',
          );
        }
        clauses += walkField(key, value, steps, depth);
      }
    }
  }
  return clauses;
}

/** One field's clause: a bare value means equality, an object means operators. */
function walkField(field, value, steps, depth) {
  if (Array.isArray(value)) {
    // Equality with the array, or membership in it? Both readings are plausible, so neither is
    // guessed at.
    throw fail(
      `field ${JSON.stringify(field)}: an array is not a filter value; use { $contains: … } for ` +
        'array membership or { $in: [ … ] } for a set of alternatives',
    );
  }
  if (value !== null && typeof value === 'object') {
    return walkOperators(field, value, steps, depth);
  }
  pushCompare(steps, field, OP.eq, value);
  return 1;
}

function walkOperators(field, operators, steps, depth) {
  if (depth > MAX_DEPTH) throw fail(`filter nested deeper than ${MAX_DEPTH}`);
  let clauses = 0;

  for (const [op, raw] of Object.entries(operators)) {
    if (raw === undefined) continue;
    switch (op) {
      case '$eq':
        pushCompare(steps, field, OP.eq, raw);
        break;
      case '$ne':
        pushCompare(steps, field, OP.ne, raw);
        break;
      case '$gt':
        pushCompare(steps, field, OP.gt, raw);
        break;
      case '$gte':
        pushCompare(steps, field, OP.gte, raw);
        break;
      case '$lt':
        pushCompare(steps, field, OP.lt, raw);
        break;
      case '$lte':
        pushCompare(steps, field, OP.lte, raw);
        break;
      case '$contains':
        pushCompare(steps, field, OP.contains, raw);
        break;
      case '$startsWith':
        if (typeof raw !== 'string') {
          throw fail(`field ${JSON.stringify(field)}: $startsWith takes a string`);
        }
        steps.push(step(STEP.compareString, { field, op: OP.startsWith, text: raw }));
        break;
      case '$in':
      case '$nin': {
        // Not in the C ABI: `$in` is an OR of equalities, and `$nin` its negation. Desugared
        // here rather than added to a frozen ABI, and the equivalence is exact because `$ne` is
        // the exact negation of `$eq` — including for documents lacking the field.
        const values = requireArray(field, op, raw);
        if (values.length === 0) throw fail(`field ${JSON.stringify(field)}: ${op} needs a value`);
        for (const value of values) pushCompare(steps, field, OP.eq, value);
        if (values.length > 1) {
          steps.push(step(STEP.combine, { op: COMBINE.or, count: values.length }));
        }
        if (op === '$nin') steps.push(step(STEP.combine, { op: COMBINE.not, count: 1 }));
        break;
      }
      case '$exists':
        steps.push(
          step(STEP.unary, { field, op: raw ? UNARY.exists : UNARY.isNull }),
        );
        break;
      default:
        throw fail(`field ${JSON.stringify(field)}: unknown operator ${JSON.stringify(op)}`);
    }
    clauses++;
  }

  // `{ field: {} }` constrains nothing, and an operator object that constrains nothing is far
  // more likely to be a typo than an intention.
  if (clauses === 0) {
    throw fail(`field ${JSON.stringify(field)}: an empty operator object matches everything`);
  }
  conjoin(steps, clauses);
  return 1;
}

/**
 * Push a comparison, choosing the step kind from the value's JavaScript type.
 *
 * `null` becomes an existence test rather than a comparison. The C ABI has no null comparison
 * because it needs none: an absent field already equals null, so `{ x: null }` is exactly
 * "x is absent or explicitly null", which is what `IS_NULL` tests.
 */
function pushCompare(steps, field, op, value) {
  if (value === null) {
    if (op === OP.eq) {
      steps.push(step(STEP.unary, { field, op: UNARY.isNull }));
      return;
    }
    if (op === OP.ne) {
      steps.push(step(STEP.unary, { field, op: UNARY.isNull }));
      steps.push(step(STEP.combine, { op: COMBINE.not, count: 1 }));
      return;
    }
    // An ordering against null is false for every document, and the engine says so — but there
    // is no way to spell it through the C ABI, and silently matching nothing would look like a
    // working query returning no results.
    throw fail(
      `field ${JSON.stringify(field)}: null can only be compared with $eq or $ne; an ordering ` +
        'against null is never true',
    );
  }

  switch (typeof value) {
    case 'string':
      steps.push(step(STEP.compareString, { field, op, text: value }));
      return;
    case 'boolean':
      steps.push(step(STEP.compareBool, { field, op, flag: value }));
      return;
    case 'number': {
      if (!Number.isFinite(value)) {
        throw fail(`field ${JSON.stringify(field)}: ${value} is not a comparable number`);
      }
      // Integral numbers become integers, matching how metadata is stored — otherwise
      // `{ count: 3 }` would compare a float against a stored integer and, although the engine
      // handles that correctly, the asymmetry would be a trap waiting for the day it does not.
      const integral = Number.isInteger(value) && Math.abs(value) < 9e15;
      steps.push(
        step(integral ? STEP.compareI64 : STEP.compareF64, { field, op, number: value }),
      );
      return;
    }
    default:
      throw fail(
        `field ${JSON.stringify(field)}: filter values must be strings, numbers, booleans or ` +
          `null; got ${typeof value}`,
      );
  }
}

function requireFilterList(key, value) {
  if (!Array.isArray(value)) throw fail(`${key} takes an array of filters`);
  return value;
}

function requireObject(key, value) {
  if (value === null || typeof value !== 'object' || Array.isArray(value)) {
    throw fail(`${key} takes a filter object`);
  }
  return value;
}

function requireArray(field, op, value) {
  if (!Array.isArray(value)) throw fail(`field ${JSON.stringify(field)}: ${op} takes an array`);
  return value;
}

/**
 * Compile a metadata object into the flat field list the bridge writes.
 *
 * Same fixed shape as a filter step, for the same reason: the JSI layer reads the same property
 * names from every element and never has to decide what one means.
 *
 * `null` is written explicitly rather than dropped. No comparison can tell an explicit null from
 * an absent field, but `{ $exists: true }` can, and a binding that dropped the key would quietly
 * change what that reports.
 */
export function compileMetadata(metadata) {
  if (metadata === undefined || metadata === null) return [];
  if (typeof metadata !== 'object' || Array.isArray(metadata)) {
    throw fail('metadata must be an object');
  }

  const fields = [];
  for (const [key, value] of Object.entries(metadata)) {
    if (value === undefined) continue;
    if (value === null) {
      fields.push({ kind: 5, key, text: '', number: 0, flag: false });
      continue;
    }
    switch (typeof value) {
      case 'string':
        fields.push({ kind: 1, key, text: value, number: 0, flag: false });
        break;
      case 'boolean':
        fields.push({ kind: 4, key, text: '', number: 0, flag: value });
        break;
      case 'number': {
        if (!Number.isFinite(value)) {
          throw fail(`metadata field ${JSON.stringify(key)}: ${value} cannot be stored`);
        }
        const integral = Number.isInteger(value) && Math.abs(value) < 9e15;
        fields.push({ kind: integral ? 2 : 3, key, text: '', number: value, flag: false });
        break;
      }
      default:
        throw fail(
          `metadata field ${JSON.stringify(key)}: values must be strings, numbers, booleans or ` +
            `null; got ${Array.isArray(value) ? 'an array' : typeof value}. Nested objects and ` +
            'arrays are not supported yet.',
        );
    }
  }
  return fields;
}

export const constants = { STEP, OP, UNARY, COMBINE, MAX_DEPTH };
